### PR TITLE
Introduce extension macros

### PR DESCRIPTION
**Pull request by EwanC**
Implement idea from https://github.com/KhronosGroup/OpenCL-Headers/issues/248 to add an version macro to the extension headers so that users can guard application code using the macro to ensure the correct APIs are being used. This uses the XML from https://github.com/KhronosGroup/OpenCL-Docs/pull/1161 which forces extension to set a revision value.

Extensions can then increment the version when they change the APIs without breaking user code, providing applications have guarded their usage.

See [CL_MAKE_VERSION](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_MAKE_VERSION) for how the macro version is defined. Note that the definitions of this macro and `cl_version` have been taken out of the  `CL_VERSION_3_0` version guard to allow them to always be used by an application. This is because there functionality is implemented in the headers and not affected in any way by the OpenCL implementation being used, and the OpenCL version that implementation supports.

### CODE REVIEW

File: CL/cl_ext.h
Comment (EwanC) [inline CL/cl_ext.h:1779]: `cl_khr_external_memory_dx` has dissapeared after copying over `cl_ext.h` wholesale from the `gen_headers.py` output - However it looks like this was intended https://github.com/KhronosGroup/OpenCL-Docs/pull/1160

File: CL/cl_ext.h
Comment (EwanC) [inline CL/cl_ext.h:1868]: `cl_khr_external_semaphore_dx_fence` has dissapeared after copying over `cl_ext.h` wholesale from the `gen_headers.py` output - However it looks like this was intended https://github.com/KhronosGroup/OpenCL-Docs/pull/1160

**Review by bashbaug — APPROVED** [review]
LGTM.

I think `cl_ext.h` is going to change slightly when/if we remove the automated formatting checks - see https://github.com/KhronosGroup/OpenCL-Headers/pull/253 - but no need to hold up this PR until the automated formatting checks are merged.  It'll just make for a few more diffs next time we re-generate headers.

Un-guarding `CL_MAKE_VERSION` means that a pre-OpenCL 3.0 application that uses `CL_MAKE_VERSION` intentionally or unintentially (say, if it meant to use `CL_MAKE_VERSION_KHR` instead) would build with newer headers and fail to build with older headers, but the risk of that happening is probably low enough that we shouldn't worry too much about it.  I don't have a better solution, anyhow.

**Comment by bashbaug** [comment]
Merging as discussed in the June 18th teleconference.

**Pull request merged**
